frontend/stmt_inferrer.py

```diff
@@ -453,18 +457,64 @@ def _infer_class_def(node, context, solver):
 
     class_attrs = solver.z3_types.instance_attributes[node.name]
     instance_type = solver.z3_types.classes[node.name]
+    class_to_funcs = solver.z3_types.class_to_funcs[node.name]
+    base_class_to_funcs = {}
+    base_attrs = {}
+    if node.bases and isinstance(node.bases[0], ast.Name):
+        base_class_to_funcs = solver.z3_types.class_to_funcs[node.bases[0].id]
+        base_attrs = solver.z3_types.instance_attributes[node.bases[0].id]
 
     for attr in class_context.types_map:
         solver.add(class_attrs[attr] == class_context.types_map[attr],
                    fail_message="Class attribute in {}".format(node.lineno))
 
-    # Set the type of the first arg in the class methods to be instance of this class
-    # TODO check staticmethod decorator
-    for func_name, args_count in solver.config.class_to_funcs[node.name]:
-        func_type = class_context.get_type(func_name)
-        arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_1".format(args_count))
-        solver.add(arg_accessor(func_type) == instance_type,
-                   fail_message="First arg in class methods has class instance type")
+        if attr in class_to_funcs and "staticmethod" not in class_to_funcs[attr][1]:
+            # If the attribute is a non static method, set the type of the first arg to be an instance of this class
+            args_len = class_to_funcs[attr][0]
+            arg_accessor = getattr(solver.z3_types.type_sort, "func_{}_arg_1".format(args_len))
+            solver.add(arg_accessor(class_attrs[attr]) == instance_type,
+                       fail_message="First arg in instance methods has class instance type")
+
+        if attr != "__init__" and attr in base_class_to_funcs:
+            # attr is an overridden method
+            # class_to_funcs[attr] is a tuple of two elements.
+            # The first is the args length, the second is the decorators list, the third is default args length
+            base_args_len = base_class_to_funcs[attr][0]
+            base_defaults_len = base_class_to_funcs[attr][2]
+            base_non_defaults_len = base_args_len - base_defaults_len
+            sub_args_len = class_to_funcs[attr][0]
+            sub_defaults_len = class_to_funcs[attr][2]
+            sub_non_defaults_len = sub_args_len - sub_defaults_len
+            decorators = base_class_to_funcs[attr][1] + class_to_funcs[attr][1]
```

Comment:
What do we generally do with decorators? I think we should reject decorators that we don't know, since they can in principle change types arbitrarily